pkg/lib/authn/otp/kind_forgot_password.go

```diff
@@ -57,35 +46,35 @@ func (k kindForgotPassword) ValidPeriod() time.Duration {
 	panic("unknown forgot password otp form")
 }
 
-func (k kindForgotPassword) RateLimitTriggerPerIP(ip string) ratelimit.BucketSpec {
-	return ratelimit.NewBucketSpec(
-		selectByChannel(
-			k.channel,
-			k.config.ForgotPassword.RateLimits.Email.TriggerPerIP,
-			k.config.ForgotPassword.RateLimits.SMS.TriggerPerIP,
-			k.config.ForgotPassword.RateLimits.SMS.TriggerPerIP,
-		),
-		selectByChannel(
-			k.channel,
-			ForgotPasswordTriggerEmailPerIP,
-			ForgotPasswordTriggerSMSPerIP,
-			ForgotPasswordTriggerWhatsappPerIP,
-		),
-		ip,
-	)
-}
-
-func (k kindForgotPassword) RateLimitTriggerPerUser(userID string) ratelimit.BucketSpec {
-	return ratelimit.BucketSpecDisabled
+func (k kindForgotPassword) RateLimitTrigger(
+	featureConfig *config.FeatureConfig,
+	envConfig *config.RateLimitsEnvironmentConfig,
+	ip string, userID string,
```

Comment:
The original code does not take userID.

Reply:
Some otp kind take userID, some does not. As all otp kind implement the same interface, I can only make it always take user ID, like originally, forgot password rate limit also implements RateLimitTriggerPerUser.

I think the responsibility of deciding whether a userID should participate in rate limit should be in the type `RateLimit`. If that rate limit does not support `perUser` or `perUserPerIP` rate limits, it will just ignore the userID and do not return related buckets. The user of the RateLimit should always provide all available information.